File: Data Structures & Algorithms/interleaving-string/submission-15.java
class Solution {
    Boolean[][] dp;
    public boolean isInterleave(String s1, String s2, String s3) {
        dp = new Boolean[s1.length()+1][s2.length()+1];
        if(s1.length()+s2.length() != s3.length())
            return false;
        return dfs(0,0,s1,s2,s3);
        

        
    }

    public boolean dfs(int i,int j,String s1,String s2,String s3){
        if(i==s1.length() && j == s2.length()){
            return true;
        }
        int k = i+j;
        if(dp[i][j]!=null)
            return dp[i][j];
        if(i<s1.length() && s1.charAt(i)==s3.charAt(k) && dfs(i+1,j,s1,s2,s3)){
            dp[i][j] = true;
            return true;
        } 
        if(j<s2.length() && s2.charAt(j)==s3.charAt(k) && dfs(i,j+1,s1,s2,s3)){
            dp[i][j] = true;
            return true;
        }
        dp[i][j] = false;
        return false;
    }
}

File: Data Structures & Algorithms/target-sum/submission-15.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        N = len(nums)
        dp = [defaultdict(int) for _ in range(N+1)]
        dp[N][target] = 1
        s = sum(nums)
        for i in range(N-1,-1,-1):
            for cur in range(s,-(s+1),-1):
                dp[i][cur] = dp[i+1][cur+nums[i]] + dp[i+1][cur-nums[i]]
        return dp[0][0]
        # def dfs(i,cur):
        #     if i == N:
        #         return 1 if cur == target else 0
        #     if (i,cur) in dp:
        #         return dp[(i,cur)]
        #     dp[(i,cur)] = dfs(i+1,cur+nums[i]) + dfs(i+1,cur-nums[i])
        #     return dp[(i,cur)]
        # return dfs(0,0)




#         2   2   2   target = 2
#                                                         0,0
#                                     2,1                                     -2,1
#             4,2(1)                          0,2(1)                  0,2(1)          -4,2
# 6,3                 2,3                 2,3         -2,3
# 0                    1
        